flocker/acceptance/testtools.py

```diff
@@ -869,24 +872,56 @@ def failed_query(failure):
                        failed_query)
         return d
     agents_connected = loop_until(reactor, nodes_available)
+    agents_connected.addCallback(lambda _: _add_nodes(cluster))
+    return agents_connected
+
+
+def _add_nodes(cluster):
+    """
+    Configure the ``Node`` objects for a newly created ``Cluster`` whose
+    nodes are known to be alive.
+
+    :param Cluster cluster: Cluster that still needs nodes set.
+    :return: ``cluster`` updated with appropriate ``nodes`` set.
+    """
+    # By default we just trust address returned by Flocker
+    def default_get_public_ip(address):
+        return address
 
-    # Extract node hostnames from API that lists nodes. Currently we
-    # happen know these in advance, but in FLOC-1631 node identification
-    # will switch to UUIDs instead.
-    agents_connected.addCallback(lambda _: cluster.current_nodes())
+    try:
+        backend = get_backend_api(None, cluster.cluster_uuid)
+    except SkipTest:
+        # Can't load backend, will have to trust Flocker's reported IPs.
+        get_public_ip = default_get_public_ip
+    else:
+        if ICloudAPI.providedBy(backend):
+            node_ips = list(set(ip_address(i) for i in ips)
+                            for ips in backend.list_live_nodes().values())
+
+            def get_public_ip(address):
+                for ips in node_ips:
+                    if ip_address(address) in ips:
+                        return [unicode(ip) for ip in ips
```

Comment:
prefer `list(unicode(ip) for ...)` rather than `[unicode(ip) ...]` so that we don't pollute the `ip` variable into the outer namespace.